### ISSUE TITLE
Not working on 5.0.2

### ISSUE THREAD

**Issue by licaon-kter** [M66B/NetGuard#20]
Tested 0.7 on a Sony 5.1.1 on a (no-root) device and looked like it worked fine at least on the great ones ( Gmail, Hangouts, Maps, Opera, YMail ).

Installing 0.7 on 5.0.2 on another (no-root) device yielded mixed results.
At first it looked like it worked on some apps ( like Opera and Orfox+Orbot ), but not on others ( Slack, Gmail, Telegram ). 

Now testing some more on the 5.0.2 device and updating to 0.8 and 0.8-proguard-2, I can't have it working for any app whatsoever.

Basically using any connection Wi-Fi or mobile:

* normal connection works ok, apps connect
* other VPN works ok, apps connect
* NetGuard active ( even NOTHING blocked ) does not work at all, apps are not connecting

Any ideas? 

/LE: IT was 0.7 and not 0.6.

**Comment by M66B** [comment]
Please open the about dialog using the action bar menu, tap seven times on the dialog and send me the logcat as proposed, so I can look into it.

Note that you cannot have two applications using the Android VPN service running at the same time.

**Comment by licaon-kter** [comment]
I can't have 2 apps running, say Gmail and Opera?

Anyway not even 1 at a time works.

**Comment by licaon-kter** [comment]
Repro steps:
* WiFi is on
* Start NetGuard, turn on, everything is NOT blocked
* Start Opera
* Try an address but nothing happens

Info:

NetGuard: 0.8/8
Android: 5.0.2 (SDK 21)

Brand: samsung
Manufacturer: samsung
Model: SM-A500FU
Product: a5ultexx
Device: a5ulte
Host: SWHD7306
Display: LRX22G.A500FUXXU1BOH3
Id: LRX22G

Logcat: https://gist.github.com/licaon-kter/c2a401a982d8999a84d5

**Comment by M66B** [comment]
I don't see anything unusual in the logcat.
My best guess is that the VPN service implementation of this device/ROM is broken.

**Comment by licaon-kter** [comment]
What type of VPN do you create OpenVPN?
My tested working one is PPTP.

**Comment by M66B** [comment]
There is not really an VPN tunnel, it is a local VPN service
See here for details: http://developer.android.com/reference/android/net/VpnService.html

**Comment by licaon-kter** [comment]
Can you point me to another app that maybe uses this? If you have one in mind for me to test?

**Comment by M66B** [comment]
I guess you have already one: "*... trying the VPN trick in Orbot did not work either*"
You can try a closed source alternative: https://play.google.com/store/apps/details?id=app.greyshirts.firewall

**Comment by licaon-kter** [comment]
Thank for that app, unfortunately *NoRoot Firewall* **WORKS** as expected, nice.

Now for this open source one, how else can I help you troubleshoot NetGuard.

**Comment by M66B** [comment]
To be sure: did you enabled NetGuard using the switch and did you block the applications you have tested? I see no reason why NetGuard shouldn't work, but NoRoot Firewall does.

**Comment by M66B** [comment]
Please install this debug version: http://www.faircode.eu/NetGuard-v0.9-debug.apk
Enable NetGuard, allow the apps you like to test, test the apps.
After that, open the about dialog using the menu, tap seven times on the about dialog and send me the debug information, referencing this issue.

**Comment by licaon-kter** [comment]
I'll test that version a bit later.

Keep in mind that with Netguard enabled on this device, apps don't connect to the internet no matter if blocked or not.

**Comment by licaon-kter** [comment]
**NetGuard-v0.9-debug.apk** starts and then it FC when I try to enable it.

**Comment by M66B** [comment]
I have just fixed this, I will build a new version.

**Comment by M66B** [comment]
This version should work again:

**Comment by licaon-kter** [comment]
**debug-2** at least it does not FC on enable, but it still does not work.

New logcat, some interesting IOExceptions: https://gist.github.com/licaon-kter/0f01e4096abca6ed1048

Also I can see ( why can't I use screenshot Samsung, why? ) that in the VPN properties it's got a transmitted/received counter, and it shows eg. 0.00B/7.7Mb, I guess that received is from the apps and transmitted is from the internet?

BTW, 0.9 works great on the 5.1.1 device.

**Comment by M66B** [comment]
Seems to block the browser and allow the rest:

```
10-30 21:38:43.405 30808 30894 I NetGuard.Service: Connect /185.26.181.220:443 uid=10195 pkg=com.opera.browser.beta
```

Correct?

Don't worry about the IOException's.

**Comment by licaon-kter** [comment]
Nope, no app can connect, I use the browser since it's pretty straight forward if it works or not.

**Comment by M66B** [comment]
Is 'com.opera.browser.beta' in '...snip plenty of Allowing' ?

**Comment by licaon-kter** [comment]
Yep they all are.

**Comment by M66B** [comment]
Then this is a bug in the ROM :-(

**Comment by M66B** [comment]
There is nothing I can do to fix this.

**Comment by licaon-kter** [comment]
Then I'll resort to use the closed source no-root firewall on this device, but I'll also test each release just in case.

Thanks for your patience.

**Comment by licaon-kter** [comment]
Tried a full reset to factory just in case.
* 0.9-debug-2 not working, I'll generate a logcat later
* 0.7 had to reboot the device as the VPN was not created on enabling the app after install, but still not working
* [noroot data firewall](https://play.google.com/store/apps/details?id=com.jianjia.firewall) works ok ( I saw you asked the dev about open sourcing it, I guess that's a reason for NetGuards existence then? :P )

**Comment by M66B** [comment]
I will explain the problem and why the other firewall works: your ROM isn't handling the function [addDisallowedApplication](http://developer.android.com/reference/android/net/VpnService.Builder.html#addDisallowedApplication(java.lang.String)). NetGuard uses this method to prevent traffic of allowed applications from being routed into the VPN sinkhole. Despite calling this function, your ROM routes all traffic into the VPN sinkhole. For the other firewall this is no problem, because the code in the VPN service decides what to block and allow (at the expense of battery usage, because all IP packets needs to be decoded, handled and encoded).

**Comment by licaon-kter** [comment]
Looks like you nailed it, [OpenVPN for Android](https://github.com/schwabe/ics-openvpn) does not work either. It connects fine, it lists the allowed apps, but no trafic is allowed.

**Comment by licaon-kter** [comment]
0.77beta2 w/ filtering ON fixes this, as expected: https://github.com/M66B/NetGuard/issues/79#issuecomment-174284829

**Comment by M66B** [comment]
Isn't that nice ;-)

**Comment by licaon-kter** [comment]
Using your build: https://github.com/M66B/NetGuard/issues/42#issuecomment-174307558

All traffic is blocked, logcat: https://gist.github.com/licaon-kter/9436637cf7d0079de3a6
